corsair-cpro.c

```diff
@@ -135,11 +147,12 @@ static int ccp_raw_event(struct hid_device *hdev, struct hid_report *report, u8
 	struct ccp_device *ccp = hid_get_drvdata(hdev);
 
 	/* only copy buffer when requested */
-	if (completion_done(&ccp->wait_input_report))
-		return 0;
-
-	memcpy(ccp->buffer, data, min(IN_BUFFER_SIZE, size));
-	complete(&ccp->wait_input_report);
+	spin_lock(&ccp->wait_input_report_lock);
+	if (!completion_done(&ccp->wait_input_report)) {
+		memcpy(ccp->buffer, data, min(IN_BUFFER_SIZE, size));
+		complete(&ccp->wait_input_report);
```

Comment:
Actually, I think `complete_all` is necessary for the `!completion_done` check to work as intended.

If only `complete` is used, then it will increment `<completion>->done`, but `wait_for_...` will immediately decrement it again. Thus, `<completion>->done` would remain `0` almost all the time, making the check (which can summed by `if (!<completion>->done)`) effectively useless.

On the other hand, `completion_all` sets the inner `done` field to a marker value (`UINT_MAX`) which isn't touched by `wait_for...`. This means that once `completion_all` is called, `completion_done` will return true until `reinit_completion` is called.

Since we want to use this check to try to only process reports that have been requested by the driver (at the very least as an optimization, but also to try to avoid clobbering `buffer` while some other part of the driver is reading it), we need the branch to not be retaken until `reinit_completion` is called again.

Reply:
Oh, yes, you are right. It is needed. Thanks.

Reply:
@jonasmalacofilho Right, thanks.

I'll see if I can create a document for liquidtux as a guide for this pattern, with this pitfall noted.

Reply:
That's a good idea!